### ISSUE TITLE
postgres.js pool support

### ISSUE THREAD

**Issue by DanielFGray** [graphile/worker#441]
### Feature description

I think Worker should ideally be able to accept a pool from either node-postgres or postgres.js

### Motivating example

I am currently working on a project that connects to the db using postgres.js, my app does not use node-postgres anywhere, and ideally I shouldn't be required to add another adapter.

### Breaking changes

I don't believe this should constitute a breaking change

### Supporting development

I [tick all that apply]:

- [x] am interested in building this feature myself
- [x] am interested in collaborating on building this feature
- [x] am willing to help testing this feature before it's released
- [x] am willing to write a test-driven test suite for this feature (before it exists)
- [x] am a [Graphile sponsor](https://www.graphile.org/sponsor/) ❤️
- [ ] have an active [support or consultancy contract](https://www.graphile.org/support/) with Graphile

**Comment by DanielFGray** [comment]
Happy to help where I can on this, though I fear my attempt wouldn't quite be up to par with the usual Graphile standards

**Comment by benjie** [comment]
Ideally we'd use an adaptor system like with PostGraphile V5 (or, more specifically, @dataplan/pg) such that we can use any client for which there is an adaptor. We have graphile-config for config, and @graphile/logger for logging, so we should probably extract the adaptors from @dataplan/pg and make @graphile/pg or similar that all our projects can use. In the mean time, we use _very few_ of the Postgres methods, so you could probably create a stub pool/client combo with anything that quacks sufficiently like Pool/Client that we'd believe that's what it was. `Pool.connect`/`Pool.end`/`Pool.on('error')`/`Pool.on('connect')` and `PoolClient.query`/`PoolClient.on('error')`/`PoolClient.on('notification')` are the main things I can think of off-hand.

**Comment by DanielFGray** [comment]
for whatever it's worth (and mostly to anyone else who might stumble on this issue), I found [postgres-bridge](https://github.com/gajus/postgres-bridge) which I expected to stub the necessary bits, but I would get either `TypeError: pgPool.query is not a function` or `TypeError: pgPool.connect is not a function` depending on if I pass the bridge instance directly as a pool, or the result of `await bridge.connect()` and I couldn't quite narrow it down any further

**Comment by lukasa1993** [comment]
any love for this ?

**Comment by leo91000** [comment]
Hello @benjie 

I'd like to try to contribute for this issue, in my case I specifically want graphile_worker to use PGLite for running my test in CI. Would you allow me to do it ?

If yes do you have any recommandation/requirements before I start ?

**Comment by benjie** [comment]
Absolutely, that would be great. I can’t remember off-hand if there’s anything specific we need to keep in mind, have a careful look over the @dataplan/pg adaptors and see if there’s anything funky going on. You’ll need to implement the listener interface too, that will likely involve a bit more work, and for Worker error recovery is critical so there will be a very high bar to getting this through. It may make sense to do it in multiple phases, but even a proof of concept would help advance this. Please note that main current contains unreleased code.

**Comment by leo91000** [comment]
Hi @benjie,

I've been digging deep into how various PostgreSQL providers work (node-postgres, postgres.js, PGLite, Neon, Vercel) to understand what a proper adapter interface would need. Here's what I've found and my proposal:

**Proposed Interface**

Based on my research, here's an extended version of what I came up with:

```typescript
export interface PgQueryable {
  /**
   * Stream-first query method
   */
  query<T extends MaybeRow = any>(
    sql: string,
    params?: any[],
  ): PgQueryResult<T>;
}

export interface PgConnection extends PgQueryable {
  /**
   * Transaction support - critical for providers without connection pooling
   */
  transaction<T>(callback: (tx: PgQueryable) => Promise<T>): Promise<T>;

  /**
   * Get a client from the pool (optional for single-connection providers)
   */
  reserve(): Promise<PgClient>;

  /**
   * LISTEN support with provider-specific implementations
   */
  listen(
    channel: string,
    onnotify: (payload: string | null) => void,
  ): Promise<{ unlisten: () => void }>;

  /**
   * Properly close the connection/pool
   */
  end(): Promise<void>;
}

export interface PgClient extends PgQueryable {
  /**
   * Release the client back to the pool
   */
  release(): void;
}
```

## Challenges

### 1. **LISTEN/NOTIFY Implementation Varies Significantly**

Each provider handles connection failures differently:

- **node-postgres**: No built-in reconnection - requires manual implementation (like Worker currently does)
- **postgres.js**: Automatic reconnection built-in, but uses simple retry (not exponential backoff like Worker)
- **PGLite**: No connection to lose (embedded), just internal event dispatching
- **Neon/Vercel**: Same as node-postgres

The adapter interface needs to abstract these differences while allowing Worker to maintain its robust exponential backoff retry strategy, and I'm not sure how to do that. If you have any thoughts on that, please share!
Do you think we should let each provider handle reconnection on their own?

### 2. **Transaction API is Essential**

Different providers have different transaction patterns:

- **node-postgres**: Manual BEGIN/COMMIT with connection from pool
- **postgres.js**: `sql.begin(callback)` with automatic management, but we could use manual BEGIN/COMMIT with reservation of client, so it's not as critical
- **PGLite**: No connection pool - uses mutexes internally to ensure only one transaction at a time (I'm not 100% sure this is correct, but it's what I've seen reading the source)

PGLite specifically needs the transaction API because:

```typescript
// PGLite uses internal mutex for serialization
await pg.transaction(async (tx) => {
  // All operations serialized via mutex
});
```

Without using their transaction API, we'd bypass their internal coordination mechanism.

In `graphile/worker`, this is OK because we don't use transactions, but if I understand correctly, the `graphile/crystal` does use transactions, but it's behind withTransaction(), so it will not be hard to migrate.

### 3. **Query Without Pool Reservation**

Some providers benefit from not requiring pool reservation:

- **Neon**: Can use HTTP for one-off queries (much faster for single queries), and use websocket when we `connect()`
- **PGLite**: Single connection, no pool to reserve from
- **postgres.js**: Automatic pool management

This is why the main `query()` method should work without explicit `reserve()`.

Maybe in later phase we should check if some queries in both repos can benefit from one off queries?

### 4. **Reserve/Release API Challenges**

For PGLite, implementing `reserve()`/`release()` is possible but semantically odd:

- There's no pool to reserve from (single embedded instance)
- Could implement something that just returns simple query, but it's not really "reserved".


What is your opinion on this proposition? Do you think it's a good idea?
Also is this issue the right way to ask questions ? Do you prefer other platform such as discord ?

**Comment by benjie** [comment]
> Based on my research, here's an extended version of what I came up with:

Is there a particular reason the `WithPgClient`/`PgClient` abstraction `@dataplan/pg` uses would not work? I'd rather avoid rewriting PostGraphile unless there's a really good reason! I designed that interface with the intention of it being as compatible as possible, and I think it still works for the database adaptors you suggest.

> The adapter interface needs to abstract these differences while allowing Worker to maintain its robust exponential backoff retry strategy, and I'm not sure how to do that.

Disconnection from the `LISTEN`/`NOTIFY` channel can (should?) be handled by the adaptor. I think it's handled by the adaptor in @dataplan/pg, but don't recall for sure.

> the graphile/crystal does use transactions, but it's behind withTransaction(), so it will not be hard to migrate.

Yes, the `pgClient.withTransaction(...)` approach that @dataplan/pg uses should work for all of these patterns, and you can add a top-level method to `WithPgClient` (`WithPgClient` is both a function and an object - it already has a `.release()` method) so you can do e.g. `context.withPgClient.tx(client => ...)`

>  No connection pool - uses mutexes internally to ensure only one transaction at a time

🤯 I had no idea PGLite was serialized like that; I wonder why they made that decision - seems somewhat disappointing.

> Query Without Pool Reservation

Adding an extra method for this makes sense; e.g. `context.withPgClient.query(...)`

> For PGLite, implementing reserve()/release() is possible but semantically odd:

Does it allow individual queries to build up a state (`SET ...;` followed by `SELECT ... current_setting(...)...`)? If not, everything in PGLite for PostGraphile would need to be in transactions anyway.

>  Also is this issue the right way to ask questions? Do you prefer other platform such as discord?

This is better for long-form discussions with code. If you want help with a variable name, Discord is probably better 😉

**Comment by benjie** [comment]
> Yes this is disapointing, I could find this as explanation why [electric-sql/pglite#how-it-works](https://github.com/electric-sql/pglite?tab=readme-ov-file&rgh-link-date=2025-07-07T10%3A56%3A10.000Z#how-it-works)

Ah; super interesting! Makes complete sense. Also explains why it would work for chained queries - effectively it's one "terminal session" that we're writing into.
